fix: Add functionality for creating recurring events

meetulr opened the pull request:
**What kind of change does this PR introduce?**
It adds the functionality for creating recurring events based on their recurrence patterns.

**Issue Number:**

Fixes #1583

**Did you add tests for your changes?**

Yes

**Does this PR introduces any breaking changes?**

No

**Summary**

To implement creation and management of recurring events, we're following [this](https://meetulr.github.io/Recurrence-Plan/) approach.
This PR implements the first part of it. i.e. creation of recurring events.

**Additional Context**

This is the first in the series of PRs for this issue, after this we would be adding the functionalities for querying, updating and deleting recurring events.

meetulr commented:
No changes were made on the frontend regarding this, so by default it creates a weekly recurring event:

https://github.com/PalisadoesFoundation/talawa-api/assets/55585268/f9105177-1b55-4fd6-962e-b2f04263ae54


We could create custom recurrence patterns:

https://github.com/PalisadoesFoundation/talawa-api/assets/55585268/5c7cebbe-5b1c-4f5b-ab17-fbb2d0c11401


https://github.com/PalisadoesFoundation/talawa-api/assets/55585268/124871cd-bcf1-4e1d-a571-f9439044cfe1

meetulr commented:
@palisadoes The number of files are large because I moved helper functions into their separate files, for ease of understanding and use later on. And also removed some of the files that were merged with previous PRs for this issue.
This PR is about creating recurring events. I've structured the code this way to make things easier for further steps (query, updation, deletion).
So, subsequent PRs won't contain these many files. This PR sort of sets things up for future work.

Cioppolo14 commented:
@meetulr Please fix the failing tests.

meetulr commented:
The test if failing because there are formatting issues in many files (files not associated with this PR), and it's running the check on all of them. Due to this other tests are skipped too. I think it's happening since the testing behaviour changed #1799. It's running the test on all the files with ".ts" extension.
The files in this PR are already linted by the pre-commit hook.

palisadoes commented:
Please do a minor commit. I made a change that hopefully should fix it.

meetulr commented:
This is ready for review.

SiddheshKukade commented on src/helpers/event/createEventHelpers/createRecurringEvent.ts:
```diff
@@ -49,20 +48,14 @@ export const createRecurringEvent = async (
-        startDate: convertedStartDate,
-        endDate: convertedEndDate,
+        startDate: data.startDate,
+        endDate: data.endDate,
```
are you assuming that sender would send the date in UTC format ? or should we perform the convertToUTCDate() here?

meetulr replied:
Yes, the dates are formatted by already by dayjs, we're recieving dates in UTC. That's why I removed the unnecessary date formatting. I've been testing this for past couple of days.

SiddheshKukade commented on src/helpers/event/recurringEventHelpers/generateRecurringEventInstances.ts:
```diff
@@ -51,12 +50,10 @@ export const generateRecurringEventInstances = async ({
-      startDate: formattedInstanceDate,
-      endDate: formattedInstanceDate,
+      startDate: date,
+      endDate: date,
```
Also make sure abut date formatting here

meetulr replied:
same here, rrule returns the dates in proper format, no need to format it again.

SiddheshKukade commented:
@meetulr can you provide answer to my questions so that I can review again ?

meetulr commented:
> @meetulr can you provide answer to my questions so that I can review again ?

I did in the comments. I'm sorry I shouldn't have resolved it so soon😅.

> are you assuming that sender would send the date in UTC format ? or should we perform the convertToUTCDate() here?

Yes, the dates are already formatted by dayjs on the frontend, and by graphQL, we're receiving dates in UTC midnight in the backend. I didn't realize that before adding the convertToUTCDate and timeZoneOffset utils, and formatting dates using date-fns's format... causing unnecessary confusion.
That wasn't required, so I removed it later. And that's why I've converted the dates to UTC midnight in the tests to simulate this behavior.
I've been testing this for past couple of days.

> Also make sure abut date formatting here

Same here, rrule returns the dates in proper format, no need to format it again. It would only serve to cause timezone discrepencies and incorrect date output.

EshaanAgg reviewed (approved):
LGTM. Would just advice you to add the documentation about all of the logic of this recurrence somewhere. Also you need to now create new events based on the queries that come in, so please take a look into that as well.

meetulr commented:
> LGTM. Would just advice you to add the documentation about all of the logic of this recurrence somewhere. Also you need to now create new events based on the queries that come in, so please take a look into that as well.

Yes, I have added docs and comments in the files. Would add some more if required.

The next PR (for query i.e. dynamic generation) is almost ready. Just have to write the tests.
